src/pow.cpp

```diff
@@ -20,34 +20,65 @@ unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHead
         return nProofOfWorkLimit;
 
     // Only change once per difficulty adjustment interval
-    if ((pindexLast->nHeight+1) % params.DifficultyAdjustmentInterval() != 0)
-    {
-        if (params.fPowAllowMinDifficultyBlocks)
-        {
-            // Special difficulty rule for testnet:
-            // If the new block's timestamp is more than 2* 10 minutes
-            // then allow mining of a min-difficulty block.
-            if (pblock->GetBlockTime() > pindexLast->GetBlockTime() + params.nPowTargetSpacing*2)
-                return nProofOfWorkLimit;
-            else
-            {
-                // Return the last non-special-min-difficulty-rules-block
-                const CBlockIndex* pindex = pindexLast;
-                while (pindex->pprev && pindex->nHeight % params.DifficultyAdjustmentInterval() != 0 && pindex->nBits == nProofOfWorkLimit)
-                    pindex = pindex->pprev;
-                return pindex->nBits;
-            }
+    uint32_t nHeight = pindexLast->nHeight + 1;
+    if (nHeight % params.DifficultyAdjustmentInterval() == 0) {
+        // Go back by what we want to be 14 days worth of blocks
+        assert(nHeight >= params.DifficultyAdjustmentInterval());
+        uint32_t nHeightFirst = nHeight - params.DifficultyAdjustmentInterval();
+        const CBlockIndex *pindexFirst = pindexLast->GetAncestor(nHeightFirst);
+        assert(pindexFirst);
+
+        return CalculateNextWorkRequired(pindexLast,
+                                         pindexFirst->GetBlockTime(), params);
+    }
+
+    if (params.fPowAllowMinDifficultyBlocks) {
+        // Special difficulty rule for testnet:
+        // If the new block's timestamp is more than 2* 10 minutes then allow
+        // mining of a min-difficulty block.
+        if (pblock->GetBlockTime() >
+            pindexLast->GetBlockTime() + 2 * params.nPowTargetSpacing) {
+            return nProofOfWorkLimit;
         }
-        return pindexLast->nBits;
+
+        // Return the last non-special-min-difficulty-rules-block
+        const CBlockIndex *pindex = pindexLast;
+        while (pindex->pprev &&
+               pindex->nHeight % params.DifficultyAdjustmentInterval() != 0 &&
+               pindex->nBits == nProofOfWorkLimit) {
+            pindex = pindex->pprev;
+        }
+
+        return pindex->nBits;
     }
 
-    // Go back by what we want to be 14 days worth of blocks
-    int nHeightFirst = pindexLast->nHeight - (params.DifficultyAdjustmentInterval()-1);
-    assert(nHeightFirst >= 0);
-    const CBlockIndex* pindexFirst = pindexLast->GetAncestor(nHeightFirst);
-    assert(pindexFirst);
+    // Do not go lower than the POW limit.
+    uint32_t nBits = pindexLast->nBits;
+    if (nBits == nProofOfWorkLimit) {
+        return nProofOfWorkLimit;
```

Comment:
This is an just an early bail out. It's not expected that is can be lower than the minimum considering it comes from a existing block.

If you look after this, the code figure out if we match the condition under which the difficulty is lowered, lowers it is so and then check for minimum PoW again.